Update telemetry tracking in the email editor

costasovo opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

When we [refactored the email editor to work on top of the Editor component](https://github.com/woocommerce/woocommerce/pull/57775), we removed a bunch of custom components, and with them, we also removed some telemetry tracking (recordEvent and recordEventOnce calls).

This PR:
- reintroduces most of the tracking calls we removed,
- adds a few new,
- adds an MD file that documents currently tracked events.

#### New tracking mechanisms
We can no longer track all the events by adding tracking code directly to the component code. This PR introduces two new tracking mechanisms we can use to further extend the tracking. 

1) Delegated tracking on `document` - currently, we listen to all `click` events, and if the target matches some of the selectors we are interested in we track. This can be found in `packages/js/email-editor/src/events/dom-tracking.ts`

2) Listening to store actions - we wrap selected actions in selected stores and apartfrom triggering the original action we also perform some tracking. This can be found in `packages/js/email-editor/src/events/store-tracking.ts`

#### Summary of changes in event tracking
##### Changed events
 - `header_inserter_sidebar_closed, header_inserter_sidebar_opened` - Changed to `header_inserter_sidebar_clicked`, with `{ isOpened }` tracked in data. This change is to be aligned with other similar events we track in the header.
 - `header_listview_sidebar_closed, header_listview_sidebar_opnened` -  - Changed to `header_inserter_listview_clicked`, with `{ isOpened }` tracked in data. This change is to be aligned with other similar events we track in the header.
 - `header_save_all_button_clicked` - Renamed to `header_save_button_clicked`. We no longer have Save all button.
 
##### Newly added events
- `header_preview_dropdown_tablet_selected` - for newly added tablet preview
- `distraction_free_toggle` - for newly added option in the more menu
- `inserter_sidebar_library_pattern_selected` - we track also patterns added from the inserter
- `quick_inserter_library_block_selected` - track inserted block from quick inserter
- `quick_inserter_library_pattern_selected`  - track inserted patterns from quick inserter
- `other_inserter_library_block_selected` - track inserted blocks from other inserters we can't/don't specifically detect (e.g from command bar)
- `other_inserter_pattern_selected` - track inserted patterns from other inserters we can't/don't specifically detect (e.g, from command menu)
- `command_menu_opened` - Tracked when the command menu is opened via click or a shortcut 
- `command_menu_closed` - Tracked when the command menu is closed 

###### Removed events
- `header_campaign_name_email_title_clicked` - The header edit title component is no longer present
- `header_campaign_name_toggle_icon_clicked` - The header edit title component is no longer present
- `header_more_menu_restore_from_trash_button_clicked` - We no longer display the trash button in the more menu
- `header_more_menu_move_to_trash_button_clicked` - We no longer display the trash button in the more menu
- `keyboard_shortcuts_toggle_fullscreen` - tracked as `header_more_menu_fullscreen_mode_toggle`
- `trash_modal_cancel_button_clicked` - not restored due to being too complex to restore
- `edit_template_blocks_notification_cancel_button_clicked` - not restored due to being too complex to restore
- `keyboard_shortcuts_save` - not restored due to being too complex to restore
- `edit_template_blocks_notification_none_admin_role_notice_closed` - After the refactor, we no longer show the modal to non-admin users.



Closes WOOPLUG-3953 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. You need to enable the email editor in WooCommerce > Settings > Advanced > Features
2. You also need to enable tracking and log it to console. This can be done by adding:
```js
// enable email editor event tracking
addFilter(
	'woocommerce_email_editor_events_tracking_enabled',
	NAME_SPACE,
	() => true
);
// log events to dev console
addAction( 'woocommerce_email_editor_events', NAME_SPACE, ( editorEvents ) => {
	const { name, ...data } = editorEvents;
	console.log( '--Email editor event:', name, data );
} );
```
to `plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/index.ts`
3. You can perform actions in the editor and observe events being logged in the console.

Note that some events (e.g. template selection modal) are logged only for newly created emails so you may need to enable emals in the WP Admin menu. To do so, add following
```php
'public'       => false,
'show_ui'      => true,
'show_in_menu' => true,
'show_in_rest' => true,
```
to 

https://github.com/woocommerce/woocommerce/blob/1817abf09affb7e33fb0c4d68cdecf6a810536c4/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L122-L139	

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

costasovo committed:
Restore tracking of editor_layout_loaded event

WOOPLUG-3953

costasovo committed:
Add a mechanism for tracking actions on stores and track setDeviceType

WOOPLUG-3953

costasovo committed:
Restore header_preview_dropdown_send_test_email_selected event

WOOPLUG-3953

costasovo committed:
Add a simple mechanism for tracking clicks and track open preview

WOOPLUG-3953

costasovo committed:
Restor tracking of moving post to trash

WOOPLUG-3953

costasovo committed:
Improve naming and add commentes to store actions tracking code

WOOPLUG-3953

costasovo committed:
Restore tracking of inserter and listview sidebars

WOOPLUG-3953

costasovo committed:
Restore tracking of header_blocks_tool_button_clicked

WOOPLUG-3953

costasovo committed:
Restore tracking of header_more_menu_dropdown_toggle

WOOPLUG-3953

costasovo committed:
Add tracking of feature toggled in more menu

WOOPLUG-3953

costasovo committed:
Restore trackig of header_save_all_button_clicked event

WOOPLUG-3953

costasovo committed:
Restore tracking of header_save_email_button_clicked

WOOPLUG-3953

costasovo committed:
Restore tracking of closing inserter sidebar

WOOPLUG-3953

costasovo committed:
Add tracking of inserting blocks and patterns

WOOPLUG-3953

costasovo committed:
Restore tracking of header_preview_dropdown_clicked

WOOPLUG-3953

costasovo committed:
Reintroduce Email,Block tabs tracking in the sidebar

WOOPLUG-3953

costasovo committed:
Refactor list view, open, close tracking, and inserter open tracking

This commit unifies this event with the rest of the events that track
isOpened within the event data instead of tracking separate events
for open and close.

It also moves the tracking to the DOM to track only interactions
via the buttons in the header. The panels are sometimes programmatically closed
for example, when a user selects distraction-free mode.

WOOPLUG-3953

costasovo committed:
Add e2e test to verify selectors used for tracking are present

WOOPLUG-3953

costasovo committed:
Fix email editor page render PHP error in PHP 7.4

PHP 7.4 doesn't support array spread.
error: Uncaught Error: Cannot unpack array with string keys
WOOPLUG-3953

costasovo committed:
Update name of save email button event

It is no longes save all button but just save button
WOOPLUG-3953

costasovo committed:
Add documentation page for telemetry tracking

WOOPLUG-3953

costasovo committed:
Add tracking for command menu

WOOPLUG-3953

costasovo committed:
Add changelog

WOOPLUG-3953

triple0t reviewed (commented):
The change makes sense to me.
But when I enable tracking, "Preview in new tab" leads to an error page `localhost:8888/wp-admin/undefined`

Also, what do you think about tracking clicks on "Show template"?

It would also be nice if we could track these new additions to the Editor UI

Email Status:
<img width="266" alt="Screenshot 2025-05-28 at 11 57 18 AM" src="https://github.com/user-attachments/assets/624411f2-591b-4598-a0f5-9c81397c82b6" />

Recipients settings:
<img width="275" alt="Screenshot 2025-05-28 at 11 57 13 AM" src="https://github.com/user-attachments/assets/12d12e10-4177-4bd8-a07e-6b781f1624d8" />

Sender Options
<img width="268" alt="Screenshot 2025-05-28 at 11 59 16 AM" src="https://github.com/user-attachments/assets/8b36bc02-9bcc-406a-affd-ee2f249c7762" />

costasovo committed:
Fix the broken open preview to a new tab

This could potentialy cause multiple issues for cases where
code relied on the returned value of an action.
WOOPLUG-3953

costasovo committed:
Add tracking of the show template toggle in the preview menu

WOOPLUG-3953

costasovo committed:
Add tracking of email status change

Currently, the email editor is still not fully refactored to a module
that exports functions, so I followed the principle used for RichTextButton
and passed tracking functionality via a filter.

I consider this a temporary solution until we add exports to the email editor package.
WOOPLUG-3953

costasovo committed:
Add tracking of recipients settings in the editor integration

WOOPLUG-3953

costasovo committed:
Add tracking of Sender options in template mode's sidebar

WOOPLUG-3953

costasovo commented:
@triple0t Thank you for the review!

> But when I enable tracking, "Preview in new tab" leads to an error page localhost:8888/wp-admin/undefined

Nice catch! This pointed me to an issue that could cause other bugs. Fixed in https://github.com/woocommerce/woocommerce/pull/58294/commits/6595b640a62689a9ab705ef835c30d872f83b9b6

> Also, what do you think about tracking clicks on "Show template"?

Added in https://github.com/woocommerce/woocommerce/pull/58294/commits/88e4d6dec49905126cb75ce9da612f3e6af0ee84

> It would also be nice if we could track these new additions to the Editor UI

Added in https://github.com/woocommerce/woocommerce/pull/58294/commits/9d67a3d86aed453bbe0520f25ddf54c810f3724b
and in https://github.com/woocommerce/woocommerce/pull/58294/commits/5a3c5226881245cb81d4b1f7f5dbc8ea5ba7a628

Note: The tracking added in the last two commits happens in the integration code. We currently don't export anything from the email editor package, so I used the same mechanism as we currently use for passing the RichTextButton component. I think we should prioritize finishing the email editor JS package. It should have exports and export functions like `initializeEditor`, tracking helpers, and some components we want to offer to integrators. With the exports, we could then replace passing stuff via filters.

Note 2: I also chose not to add the events from the integration to the document with a list of events, as they are not part of the core email editor.

triple0t reviewed (approved):
Looks good to me.

I can confirm the events are logged.
![Screenshot 2025-05-30 at 8 36 16 AM](https://github.com/user-attachments/assets/5f70bfbd-9c47-470b-83d7-bc7782abe148)

Thank you.
